### PR TITLE
CLN: Refactor and type-Annotate time data functions

### PR DESCRIPTION
**Pull request by janbjorge**
Addresses: https://github.com/equinor/fmu-dataio/issues/457

### CODE REVIEW

File: src/fmu/dataio/_objectdata_provider.py [hunk 1]
```diff
@@ -117,7 +117,33 @@ def npfloat_to_float(v: Any) -> Any:
 
 
 @dataclass
-class _ObjectDataProvider:
```
Comment (janbjorge) [inline src/fmu/dataio/_objectdata_provider.py:120]: Module is marked as private.

File: src/fmu/dataio/_objectdata_provider.py [hunk 2]
```diff
@@ -635,59 +659,46 @@ def _check_index(self, index: list[str]) -> None:
...
-                tresult["time"].append(xfield1)
-                tresult["time"].append(xfield2)
-            else:
-                tresult["time"].append(xfield2)
-                tresult["time"].append(xfield1)
+            start, stop = (
+                TimedataValueLabel.from_list(tdata[0]),
+                TimedataValueLabel.from_list(tdata[1]),
+            )
+            if start.value > stop.value:
```
Comment (janbjorge) [inline src/fmu/dataio/_objectdata_provider.py:692]: This is flaky, works as long as no timezone is included.


```python
from datetime import datetime

from hypothesis import given
from hypothesis.strategies import datetimes, timezones


@given(a=datetimes(timezones=timezones()), b=datetimes(timezones=timezones()))
def test_iso_string(a: datetime, b: datetime) -> None:
    assert (a > b) == (a.isoformat() > b.isoformat())
```

**Review by mferrera — APPROVED** [review]
LGTM but I would suggest `TimedataFormat` and `TimedataLegacyFormat`, with the current implementation in `_derive_timedata` (i.e. getting rid of `_derive_timedata_modern/newformat` and a short circuit to the old one.

The first implies the current and up to date version while the second is the polite way to say "old". Calling something modern in code never ages well.

**Comment by perolavsvendsen** [comment]
Is this related to actual time stamps, or "FMU timestamps" (simulation time)? In case of the latter, there is a specific format. And there should not be e.g. timezones or anything like that.

**Comment by janbjorge** [comment]
> Is this related to actual time stamps, or "FMU timestamps" (simulation time)? In case of the latter, there is a specific format. And there should not be e.g. timezones or anything like that.

The timestamps are given as `%Y%m%d` and prob. wont ever contain timezone info. But i do not like the idea of doing timestamp checking on strings. Thus ive added the parsing logic.

**Comment by perolavsvendsen** [comment]
> > Is this related to actual time stamps, or "FMU timestamps" (simulation time)? In case of the latter, there is a specific format. And there should not be e.g. timezones or anything like that.
> 
> The timestamps are given as `%Y%m%d` and prob. wont ever contain timezone info. But i do not like the idea of doing timestamp checking on strings. Thus ive added the parsing logic.

👍 

I guess calling these "timestamps" is a bit of a stretch. We have discussed before if they should be proper (ISO) times. The current standard has weaknesses, for instance that it will not work for timesteps within a single day.

**Pull request merged**